### PR TITLE
Adding a few introductory tutorials

### PR DESCRIPTION
**Pull request by ktangsali**
# Modulus Pull Request

## Description

This PR adds a few beginner level tutorials. The test scripts can also be run as regression/integration tests and can potentially become part of CI. This can currently be executed by running

```
pytest test_runner.py
```

The sample output can be viewed here: https://drive.google.com/drive/folders/16Qbi43r5nrALooUGh6V18ep5a2g7pBHj?usp=drive_link (only accessible to NVIDIA employees)

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: docs/test_scripts/test_basic.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    padding=5,
+).to("cuda")
+
+optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+scheduler = torch.optim.lr_scheduler.LambdaLR(
+    optimizer, lr_lambda=lambda step: 0.85**step
+)
+
+# run for 20 iterations
+for i in range(21):
```
Comment (akshaysubr) [inline docs/test_scripts/test_basic.py:37]: This is 21 iterations

File: docs/test_scripts/test_basic_checkpointing.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    "./checkpoints",
+    models=model,
+    optimizer=optimizer,
+    scheduler=scheduler,
+    device="cuda",
+)
+
+# run for 20 iterations
+# starting with the loaded epoch
+for i in range(max(1, loaded_epoch), 21):
```
Comment (akshaysubr) [inline docs/test_scripts/test_basic_checkpointing.py:49]: Make distinction between epoch and iteration

File: docs/test_scripts/test_basic_inference.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+).to("cuda")
+
+# Save the checkpoint. For demo, we will just save untrained checkpoint, 
+# but in typical workflows is saved after model training. 
+model.save("untrained_checkpoint.mdlus")
+
+# Inference code
+
+# The parameters to instantitate the model will be loaded from the checkpoint
+model_inf = FNO.from_checkpoint("untrained_checkpoint.mdlus").to("cuda")
```
Comment (akshaysubr) [inline docs/test_scripts/test_basic_inference.py:32]: Can you change this to `Module.from_checkpoint` so you don't need to know the exact class needed at inference time? Just need the .mdlus checkpoint and that's it.

File: docs/test_scripts/test_basic_inference.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+# put the model in evaluation mode
+model_inf.eval()
+
+# run inference
+with torch.no_grad():
```
Comment (akshaysubr) [inline docs/test_scripts/test_basic_inference.py:38]: Should maybe use `torch.inference_mode()` here instead. https://pytorch.org/docs/stable/generated/torch.inference_mode.html

File: docs/test_scripts/test_console_logger.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+# Initialize the logger
+logger = PythonLogger("main")  # General python logger
+LaunchLogger.initialize()
+
+# Use logger methods to track various information during training
+logger.info("Starting Training!")
+# run for 20 iterations
+for i in range(21):
+    # wrap the epoch in launch logger to control frequency of output for console logs
+    with LaunchLogger("train", epoch=i) as log:
```
Comment (akshaysubr) [inline docs/test_scripts/test_console_logger.py:46]: Again, make clear distinction between epoch and iterations

File: docs/test_scripts/test_custom_model_demo_1.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    cuda_graph_warmup=11,
+)
+def training_step(invar, outvar):
+    predvar = model(invar)
+    loss = mse(predvar, outvar)
+    return loss
+
+
+# run for 20 iterations
+for i in range(21):
```
Comment (akshaysubr) [inline docs/test_scripts/test_custom_model_demo_1.py:140]: This is 21 iterations, not 20 like the comment says

File: docs/test_scripts/test_custom_model_demo_1.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+)
+model = MdlsUNet().to("cuda")
+
+optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+scheduler = torch.optim.lr_scheduler.LambdaLR(
+    optimizer, lr_lambda=lambda step: 0.85**step
+)
+
+
+# Create training step function with optimization wrapper
```
Comment (akshaysubr) [inline docs/test_scripts/test_custom_model_demo_1.py:127]: Add comment here saying this also runs the backward on the loss and the optimizer step as well as any mixed precision optimizations and cuda graphs

File: docs/test_scripts/test_mlflow_logger.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    mode="offline",
+)
+LaunchLogger.initialize(use_mlflow=True)
+
+# Use logger methods to track various information during training
+logger.info("Starting Training!")
+# run for 20 iterations
+for i in range(21):
+    # wrap the epoch in launch logger to control frequency of output for console logs
+    with LaunchLogger("train", epoch=i) as log:
```
Comment (akshaysubr) [inline docs/test_scripts/test_mlflow_logger.py:56]: Rename to something like `launchlog` or something? `logger` and `log` here is a bit confusing

**Review by akshaysubr — CHANGES_REQUESTED** [review]

File: docs/tutorials/simple_logging_and_checkpointing.rst [hunk 9]
```diff
@@ -0,0 +1,238 @@
+Simple Logging and Checkpointing recipe in Modulus
```
Comment (akshaysubr) [inline docs/tutorials/simple_logging_and_checkpointing.rst:1]: Also, would be good to remove `in Modulus` in all these titles. This is the Modulus documentation anyway, so this is redundant and kinda distracting.

File: docs/test_scripts/test_basic.py (hunk 1 above)
Comment (ktangsali) [reply docs/test_scripts/test_basic.py]: Thanks, fixed

File: docs/test_scripts/test_basic_checkpointing.py (hunk 2 above)
Comment (ktangsali) [reply docs/test_scripts/test_basic_checkpointing.py]: Done!

File: docs/test_scripts/test_basic_inference.py (hunk 3 above)
Comment (ktangsali) [reply docs/test_scripts/test_basic_inference.py]: Good catch, done

File: docs/test_scripts/test_custom_model_demo_1.py (hunk 7 above)
Comment (ktangsali) [reply docs/test_scripts/test_custom_model_demo_1.py:129]: Thanks! Done

File: docs/test_scripts/test_mlflow_logger.py (hunk 8 above)
Comment (ktangsali) [reply docs/test_scripts/test_mlflow_logger.py]: Sounds good. Done

File: docs/tutorials/simple_logging_and_checkpointing.rst (hunk 9 above)
Comment (ktangsali) [reply docs/tutorials/simple_logging_and_checkpointing.rst]: Removed

File: docs/test_scripts/test_simple_distributed.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        padding=5,
+    ).to("cuda")
+
+    if dist.distributed:
+        ddps = torch.cuda.Stream()
+        with torch.cuda.stream(ddps):
+            model = DistributedDataParallel(
+                model,
+                device_ids=[dist.local_rank],  # Set the device_id to be
+                # the local rank of this process on
```
Comment (akshaysubr) [inline docs/test_scripts/test_simple_distributed.py:47]: Indenting here is off

File: docs/test_scripts/test_simple_distributed.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+from modulus.utils import StaticCaptureTraining
+
+# [imports]
+
+# [code]
+
+
+def main():
+
+    DistributedManager.initialize()
```
Comment (akshaysubr) [inline docs/test_scripts/test_simple_distributed.py:19]: A comment here explaining why this is needed would be helpful

**Review by akshaysubr — APPROVED** [review]

File: docs/test_scripts/test_simple_distributed.py (hunk 10 above)
Comment (ktangsali) [reply docs/test_scripts/test_simple_distributed.py:47]: Yeah I saw this, Actually `ruff` is the culprit here, haha! Let me fix this nevertheless

Comment (ktangsali) [reply docs/test_scripts/test_simple_distributed.py]: fixed

File: docs/test_scripts/test_simple_distributed.py (hunk 11 above)
Comment (ktangsali) [reply docs/test_scripts/test_simple_distributed.py:21]: Added referring to the distributed docs.

File: docs/tutorials/simple_training_example.rst [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+.. literalinclude:: ../test_scripts/test_custom_model_demo_1.py
+   :language: python
+   :start-after: [modulus sym model] 
+   :end-before: [modulus sym model]
+
+
+Optimized Training workflow
+----------------------------
+
+Once we have a model defined in the Modulus style, we can make use the optimizations
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply docs/tutorials/simple_training_example.rst]: Thanks Oliver for the review! I have fixed the grammatical errors you pointed.

**Review by loliverhennigh — APPROVED** [review]
